File: algorithms/CheckTree/checktree.cpp
 
#include <iostream>
#include <vector>
using namespace std;

int nodes, edges, u, v;
vector<int> arr[100001];   
vector<int> vis(100001, 0); 
void dfs(int v)
{

    vis[v] = 1;
    for (auto child : arr[v])
        if (vis[child] == 0)
        {
            dfs(child);
        }
}

int main()
{
    cout << "Enter nodes and number of edges: ";
    cin >> nodes >> edges;
    cout << "Enter the edges:- \n";
    for (int i = 1; i <= edges; i++)
    {
        cin >> u >> v;
        arr[u].push_back(v);
        arr[v].push_back(u);
    }

    int cc_cnt = 0; 
    for (int i = 1; i <= nodes; i++)
    {
        if (vis[i] == 0)
        {
            dfs(i);
            cc_cnt++;
        }
    }
    if (cc_cnt == 1)
    {
        cout << "This graph is a tree\n";
    }
    else
    {
        cout << "This graph is a tree\n";
    }
    return 0;
}

File: CheckTree/checktree.py
def dfs(v, vis, arr):
    vis[v] = 1
    for child in arr[v]:
        if vis[child] == 0:
            dfs(child, vis, arr)

nodes, edges = map(int, input("Enter nodes and number of edges: ").split())
arr = [[] for _ in range(100001)]
vis = [0] * 100001

print("Enter the edges:")
for _ in range(edges):
    u, v = map(int, input().split())
    arr[u].append(v)
    arr[v].append(u)

cc_cnt = 0
for i in range(1, nodes + 1):
    if vis[i] == 0:
        dfs(i, vis, arr)
        cc_cnt += 1

if cc_cnt == 1:
    print("This graph is a tree")
else:
    print("This graph is not a tree")
